# IMPORT CÁC THƯ VIỆN CẦN THIẾT

In [1]:
import time
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import os
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB

<p> Đưa dữ liệu vào 

In [2]:
# đọc file csv 
df = pd.read_csv(r'D:\HK2-Y4\Học máy\LOAN Status\loan_data.csv')
# in ra 5 hàng đầu 
df.head()


person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49                         3.0           561   
1                 0.08                         2.0           504   
2                 0.44                         3.0           635   
3                 0.44                         2.0           675   
4                 0.53                         4.0           586   

  previous_loan_defaults_on_file  loan_status  
0                             No            1  
1                            Yes            0  
2                             No            1  
3                             No            1  
4                             No            1

In [3]:
# in ra 5 hàng cuối
df.tail()

person_age person_gender person_education  person_income  \
44995        27.0          male        Associate        47971.0   
44996        37.0        female        Associate        65800.0   
44997        33.0          male        Associate        56942.0   
44998        29.0          male         Bachelor        33164.0   
44999        24.0          male      High School        51609.0   

       person_emp_exp person_home_ownership  loan_amnt        loan_intent  \
44995               6                  RENT    15000.0            MEDICAL   
44996              17                  RENT     9000.0    HOMEIMPROVEMENT   
44997               7                  RENT     2771.0  DEBTCONSOLIDATION   
44998               4                  RENT    12000.0          EDUCATION   
44999               1                  RENT     6665.0  DEBTCONSOLIDATION   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
44995          15.66                 0.31                         3.0   
44996          14.07                 0.14                        11.0   
44997          10.02                 0.05                        10.0   
44998          13.23                 0.36                         6.0   
44999          17.05                 0.13                         3.0   

       credit_score previous_loan_defaults_on_file  loan_status  
44995           645                             No            1  
44996           621                             No            1  
44997           668                             No            1  
44998           604                             No            1  
44999           628                             No            1

# TÌM HIỂU VỀ DỮ LIỆU - EXPLORE DATA ANALYSIS(EDA)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   person_age                      45000 non-null  float64
 1   person_gender                   45000 non-null  object 
 2   person_education                45000 non-null  object 
 3   person_income                   45000 non-null  float64
 4   person_emp_exp                  45000 non-null  int64  
 5   person_home_ownership           45000 non-null  object 
 6   loan_amnt                       45000 non-null  float64
 7   loan_intent                     45000 non-null  object 
 8   loan_int_rate                   45000 non-null  float64
 9   loan_percent_income             45000 non-null  float64
 10  cb_person_cred_hist_length      45000 non-null  float64
 11  credit_score                    45000 non-null  int64  
 12  previous_loan_defaults_on_file  

In [5]:
df.isna().sum()

person_age                        0
person_gender                     0
person_education                  0
person_income                     0
person_emp_exp                    0
person_home_ownership             0
loan_amnt                         0
loan_intent                       0
loan_int_rate                     0
loan_percent_income               0
cb_person_cred_hist_length        0
credit_score                      0
previous_loan_defaults_on_file    0
loan_status                       0
dtype: int64

In [6]:
df.duplicated().sum()

0

In [7]:
df.isin([np.inf, -np.inf]).any()

person_age                        False
person_gender                     False
person_education                  False
person_income                     False
person_emp_exp                    False
person_home_ownership             False
loan_amnt                         False
loan_intent                       False
loan_int_rate                     False
loan_percent_income               False
cb_person_cred_hist_length        False
credit_score                      False
previous_loan_defaults_on_file    False
loan_status                       False
dtype: bool

**Note**:
1. Bộ dữ liệu có 14 cột và 45k dòng.
2. Bộ dữ liệu khá là sạch vì không chứa giá trị rỗng  ```null``` và giá trị trùng lặp ```duplicate```.
3. Bộ dữ liệu không chứa dạng ```infinity```.
4. Hầu hết các giá trị đều có dạng ```int``` và ```float``` trừ 5 cột :<br>
    -Cột: ```person_gender```,```person_education```,```person_home_ownership```,```loan_intent```,```previous_loan_defaults_on_file```.
    > => Cần modify kiểu dữ liệu lại các cột này để có thể huấn luyện mô hình 


In [8]:
df.describe()

person_age  person_income  person_emp_exp     loan_amnt  \
count  45000.000000   4.500000e+04    45000.000000  45000.000000   
mean      27.764178   8.031905e+04        5.410333   9583.157556   
std        6.045108   8.042250e+04        6.063532   6314.886691   
min       20.000000   8.000000e+03        0.000000    500.000000   
25%       24.000000   4.720400e+04        1.000000   5000.000000   
50%       26.000000   6.704800e+04        4.000000   8000.000000   
75%       30.000000   9.578925e+04        8.000000  12237.250000   
max      144.000000   7.200766e+06      125.000000  35000.000000   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
count   45000.000000         45000.000000                45000.000000   
mean       11.006606             0.139725                    5.867489   
std         2.978808             0.087212                    3.879702   
min         5.420000             0.000000                    2.000000   
25%         8.590000             0.070000                    3.000000   
50%        11.010000             0.120000                    4.000000   
75%        12.990000             0.190000                    8.000000   
max        20.000000             0.660000                   30.000000   

       credit_score   loan_status  
count  45000.000000  45000.000000  
mean     632.608756      0.222222  
std       50.435865      0.415744  
min      390.000000      0.000000  
25%      601.000000      0.000000  
50%      640.000000      0.000000  
75%      670.000000      0.000000  
max      850.000000      1.000000

**Note**:
1) Cột ```person_age``` có giá trị max là 144 điều này không hợp lý.
2) Cột ```person_income``` có giá trị cao nhất là 7m2 trong khi giá trị thấp nhất là 8k điều này cho thấy sự chênh lệch về mức lương.
3) Cột ```person_emp_exp``` có giá trị cao nhất là 125 có thể đây là giá trị ngoại lai vì không một ai có thể sống trên 125 tuổi để có kinh nghiệm là việc là 125.
4) Cột ```loan_amnt``` có giá trị dao động từ 500 đến 35k.
5) Cột ```loan_int_rate``` có giá trị lãi suất là 5.42% đến 20%
6) Cột ```loan_percent_income``` có giá trị là 0.66% điều này cho thấy một số người vay có nguy cơ tài chính cao.
7) Cột ```cb_person_cred_hist_length``` thời gian lịch sử trung bình là 5.87`. Tuy nhiên ta thấy có một người có lịch sử tín dụng đến 30 năm.
8) Cột ```credit_score``` có giá trị dao động từ 390 đến 850
9) Cột ```loan_status``` là biến mục tiêu nên sẽ giải thích khi khai phá thông tin các cột.


### Xử lý các trường hợp dữ liệu sai thông tin

In [9]:
# Với cột tuổi
df = df[(df['person_age'] >= 18) & (df['person_age'] <= 100)]
# Với cột kinh nghiệm việc làm 
df = df[(df['person_emp_exp'] >= 0) & (df['person_emp_exp'] <= 50)]
#in lại các thông tin thống kê của cột
df.describe()

person_age  person_income  person_emp_exp     loan_amnt  \
count  44988.000000   4.498800e+04    44988.000000  44988.000000   
mean      27.742354   7.990907e+04        5.388148   9583.323820   
std        5.881527   6.332489e+04        5.896067   6314.994983   
min       20.000000   8.000000e+03        0.000000    500.000000   
25%       24.000000   4.719425e+04        1.000000   5000.000000   
50%       26.000000   6.704550e+04        4.000000   8000.000000   
75%       30.000000   9.577900e+04        8.000000  12237.250000   
max       73.000000   2.448661e+06       50.000000  35000.000000   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
count   44988.000000         44988.000000                44988.000000   
mean       11.006634             0.139738                    5.864408   
std         2.979042             0.087209                    3.872007   
min         5.420000             0.000000                    2.000000   
25%         8.590000             0.070000                    3.000000   
50%        11.010000             0.120000                    4.000000   
75%        12.990000             0.190000                    8.000000   
max        20.000000             0.660000                   30.000000   

       credit_score   loan_status  
count  44988.000000  44988.000000  
mean     632.573308      0.222281  
std       50.389790      0.415784  
min      390.000000      0.000000  
25%      601.000000      0.000000  
50%      640.000000      0.000000  
75%      670.000000      0.000000  
max      784.000000      1.000000

In [10]:
df.shape

(44988, 14)

### Khai phá thông tin về các cột


In [11]:
# Separate numerical and categorical columns
numerical_columns = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
non_numerical_columns = df.select_dtypes(include=['object']).columns.tolist()

# Display the lists of numerical and categorical columns
print("\nNumerical Columns:", numerical_columns)
print("Categorical Columns:", non_numerical_columns)


Numerical Columns: ['person_age', 'person_income', 'person_emp_exp', 'loan_amnt', 'loan_int_rate', 'loan_percent_income', 'cb_person_cred_hist_length', 'credit_score', 'loan_status']
Categorical Columns: ['person_gender', 'person_education', 'person_home_ownership', 'loan_intent', 'previous_loan_defaults_on_file']
